src/gs-category-page.c

```diff
@@ -333,24 +333,34 @@ gs_category_page_create_filter_list (GsCategoryPage *self,
 		s = GS_CATEGORY (g_ptr_array_index (children, i));
 		if (gs_category_get_size (s) < 1) {
 			gboolean category_is_usb = FALSE;
+			gboolean apps_detected = FALSE;
 
-			g_debug ("not showing %s/%s as no apps",
-				 gs_category_get_id (category),
-				 gs_category_get_id (s));
-
-			/* re-filter USB category with no apps so the
-			 * placeholder app tiles get cleared out, then set
-			 * "empty state" message for an empty USB disk
+			/* Re-filter USB category with no apps so the
+			 * placeholder app tiles get cleared out or if it actually
+			 * detects apps on the USB
 			 */
 			if (g_strcmp0 (gs_category_get_id (category), "usb") == 0) {
-				gs_category_page_populate_filtered (self, s);
+				GPtrArray *usb_apps;
 				category_is_usb = TRUE;
+				gs_category_page_populate_filtered (self, s);
+				/* After reloading the page it will get the actual apps for the
+				 * for the usb so we test if the category will truly have the
+				 * no apps box.
+				 */
+				usb_apps = gs_category_get_children (category);
+				if ((usb_apps != NULL) && (usb_apps->len != 0))
+					apps_detected = TRUE;
+			}
+			/* Then set "empty state" message for an empty USB disk. */
+			if (!apps_detected) {
+				g_debug ("not showing %s/%s as no apps",
+					 gs_category_get_id (category),
+					 gs_category_get_id (s));
+
+				gtk_widget_set_visible (self->no_apps_box, category_is_usb);
+				gtk_widget_set_visible (self->scrolledwindow_category, !category_is_usb);
```

Comment:
I think this code would be much simpler if you rename the var `category_is_usb` as `show_no_apps_box` and just set its state, instead of having now `apps_detected` and `category_is_usb`.